Preprocess cabal-install.cabal: don't leak dev-details into production

phadej opened the pull request:
I used `m4`, but if the approach accepted generally, I can rewrite it in Haskell-only solution.

In short:
- for production the generated `cabal-install.cabal` is `cabal-version: >=1.10` and is very plain
- for development/CI we use `cabal-version: 2.0` so we can use internal libraries!

In the long run - when our support window allows - the latter version can become the default.

The only drawback, is that one have to be careful not to commit development `cabal-install.cabal` file.

23Skidoo commented:
> The only drawback, is that one have to be careful not to commit development `cabal-install.cabal` file.

Perhaps we should add `cabal-install.cabal` to `.gitignore` then? It'll still be possible to commit it with with `git add --force`. And there should obviously be a CI meta check that the generated production `cabal-install.cabal` file is identical to the one in the repo.

hvr commented:
@23Skidoo otoh, if you can have that metacheck, why would the generated .cabal file ever diverge from the one you'd generate automatically? i.e. when would you ever modify your cabal-install.cabal file w/o having modified the master one?

23Skidoo commented:
idk, people do strange things sometimes.

23Skidoo commented:
I think it's a good idea to have such checks for all generated files that are also committed to the repo.

hvr commented:
I'm all for the check; I was just wondering whether it's still needed to gitignore if you have that check

23Skidoo commented:
If it's in `.gitignore`, it doesn't clutter `git status` output and you can still run `git commit -a`

phadej commented:
I think if it's committed, than `.gitignore` doesn't apply to that file. `git` already knows about it, so to say.

23Skidoo commented:
Hmm, yes, you're right, for some reason I was sure that this worked.

23Skidoo commented:
`git update-index --assume-unchanged cabal-install/cabal-install.cabal` seems to work, however.

phadej commented:
Rewrote using DIY-preprocessor, but it's way simpler syntax in comparison to m4 (we don't need all that power). I can also split `cabal-lib-solver`, but then I'll need to move the files into separate folders.

phadej commented:
CI, was green, so I squased everything, and IMO it's mergeable.

I'm not sure if we have consensus on what to do regarding `git` and `cabal-install.cabal`, in any case that can be changed if there's some good way.

quasicomputational commented on cabal-install/cabal-install.cabal.pp:
```diff
@@ -0,0 +1,642 @@
+    Haskell software by automating the fetching, configuration, compilation
+    and installation of Haskell libraries and programs.
+    .
+    This package only provides an executable and cannot be used as a
```
Is this caveat needed any more?

phadej replied:
I think not.

23Skidoo commented:
We can make `make cabal-install-[dev,monolithic]` print a message telling the user to run `git update-index --assume-unchanged cabal-install/cabal-install.cabal` to make Git ignore changes to that file. I don't think it should do that by default.

phadej commented:
@23Skidoo good idea, I'll add that.

phadej commented on travis-script.sh:
```diff
@@ -165,17 +167,20 @@ fi
-fi
+# TODO: cabal-install doesn't have public lib, haddock doesn't make sense
+#if [ "$TRAVIS_OS_NAME" = "linux" ]; then
+#    timed cabal new-haddock cabal-install
```
I think `haddock` have to learn about (public) internal libraries properly. Or should it handle them already somehow, @alexbiehl , @hvr?

phadej replied:
i.e. I'd leave this out (and create a tracking issue),

hvr replied:
afaik, haddock gets fed internal libraries as if they were ordinary libraries by cabal; the thing we might need to adapt is to create a "TOC" maybe... but haven't thought much about it yet;

phadej commented:
I'll forward patch this to `master` tomorrow (on Monday)
